File: earthsciio/backends/zarr_write.py
"""The ``zarr`` WRITER — a streaming, sharded **Zarr v3** output backend built on
zarr-python 3.x (the Python mirror of Julia's ``julia/src/zarr_write.jl``).

streaming-output-sinks RFC, Wave 4. Where the Julia writer hand-rolls the Zarr v3
shard container byte-for-byte, this one delegates the whole encode — sharding
codec, Blosc(zstd)+shuffle inner pipeline, ``dimension_names``, growable ``time``
axis via array resize — to the mature ``zarr`` library. Conformance across
languages is **tolerance-based on decoded arrays** (RFC §16.6), never byte
identity, so using zarr-python's own container layout is explicitly allowed; the
codec *parameters* (zstd / level / shuffle / inner-chunk vs shard shape) still
match the Julia profiles so decoded output is reproducible.

Shape of the emitted store (RFC §16.1, normative):

* ``zarr_format: 3`` — one ``zarr.json`` per group and per array.
* The SHARDING codec: the array's chunk grid is the SHARD (outer, write) shape;
  the sharding codec's inner ``chunk_shape`` is the read chunk. One shard object
  packs many inner chunks (few large objects on S3/Lustre, small chunks for
  readers). Inner pipeline = bytes(little) + Blosc(zstd, shuffle).
* The ``time`` axis grows by array resize: each :meth:`ZarrWriter.write_record`
  buffers one time slice; a full time-shard (or an explicit :meth:`write_flush`)
  resizes every growable array and writes the slab.

Commit / durability. Local output goes through zarr-python's ``LocalStore``
(atomic per-object writes); ``s3://`` output through ``FsspecStore`` (s3fs). A
per-store **output manifest** (``output_manifest.json``, schema
``earthsciio/output-manifest/v1``) records committed time-shards, the last durable
``t``, the codec params, a schema fingerprint, and the base URL — the mirror of the
read cache's per-blob manifest and the Julia writer's output manifest. It is
refreshed on every flush, so a restart reads the last committed ``t`` from it.

Public surface (conceptual mirror of the Julia writer):

    w = ZarrWriter()
    h = w.write_open(base_url, schema)      # declare dims/coords/vars/chunk grid ONCE
    w.write_record(h, t, {var: slab, ...})  # persist one time record
    w.write_flush(h)                        # durable barrier (checkpoint boundary)
    w.write_close(h)                        # finalize + end-of-run manifest

``zarr``/``fsspec``/``s3fs`` are imported **lazily**, shipped as optional extras,
so a base install stays lean (matching the reader + the old ``numcodecs`` culture).
"""

from __future__ import annotations

import json
import os
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Tuple

import numpy as np

__all__ = [
    "BloscProfile",
    "BLOSC_DIAGNOSTIC",
    "BLOSC_CHECKPOINT",
    "OutputVar",
    "OutputSchema",
    "ZarrWriter",
    "ZarrWriteHandle",
    "OUTPUT_MANIFEST_SCHEMA",
]

OUTPUT_MANIFEST_SCHEMA = "earthsciio/output-manifest/v1"


# --- pinned codec profiles (RFC §16.1, matching julia/src/zarr_write.jl) ----- #


@dataclass(frozen=True)
class BloscProfile:
    """A pinned Blosc codec profile: ``cname`` / ``clevel`` / ``shuffle``."""

    cname: str
    clevel: int
    shuffle: bool


#: Diagnostic profile — Blosc **zstd** + byte-shuffle, moderate level (5).
BLOSC_DIAGNOSTIC = BloscProfile("zstd", 5, True)
#: Checkpoint profile — **lossless** Blosc zstd (level 7) + byte-shuffle.
BLOSC_CHECKPOINT = BloscProfile("zstd", 7, True)


def _profile(name: str) -> BloscProfile:
    if name == "diagnostic":
        return BLOSC_DIAGNOSTIC
    if name == "checkpoint":
        return BLOSC_CHECKPOINT
    raise ValueError(
        f"unknown codec profile {name!r} (expected 'diagnostic' or 'checkpoint')"
    )


# --- the output schema ------------------------------------------------------- #


@dataclass
class OutputVar:
    """One streaming output variable: its on-disk dim names (file order, MUST
    include the schema ``time_dim``), its numpy dtype, and optional CF variable
    attributes ``attrs`` (e.g. ``units``, ``standard_name``, and the CF
    ``coordinates`` attribute naming this variable's auxiliary coordinates). The
    attrs are written verbatim into the array node's ``attributes`` alongside the
    mechanical ``dimension_names``."""

    dims: List[str]
    dtype: Any
    attrs: Dict[str, Any] = field(default_factory=dict)

    def __post_init__(self) -> None:
        self.dims = [str(d) for d in self.dims]
        self.dtype = np.dtype(self.dtype)
        self.attrs = dict(self.attrs)


@dataclass
class OutputSchema:
    """The input to :meth:`ZarrWriter.write_open`. Mirrors the Julia
    ``OutputSchema`` (a plain struct; no EarthSciAST dependency here).

    Fields:

    * ``dims`` — ORDERED ``[(name, length), ...]``. The ``time_dim`` entry's
      length is a placeholder (0 conventional); the time axis grows.
    * ``time_dim`` — the growable axis name.
    * ``vars`` — ORDERED ``[(name, OutputVar), ...]`` streaming variables.
    * ``chunk_shape`` — ``{dim: inner-chunk length}``.
    * ``shard_shape`` — ``{dim: shard length}`` (a multiple of the inner chunk
      length along every dim). ``shard_shape[time_dim]`` = records per flushed
      shard object.
    * ``coords`` — ORDERED ``[(name, (values, attrs)), ...]`` static coordinate
      arrays, 1-D over their own dim, written once at ``write_open``. An entry
      for ``time_dim`` supplies the time coordinate's attrs (its VALUES are
      ignored — they come from the ``t`` of each record).
    * ``profile`` — ``"diagnostic"`` or ``"checkpoint"`` (codec params).
    * ``attrs`` — group-level attributes.
    * ``time_dtype`` — element type of the time coordinate (default float64).
    """

    dims: List[Tuple[str, int]]
    time_dim: str
    vars: List[Tuple[str, OutputVar]]
    chunk_shape: Dict[str, int]
    shard_shape: Dict[str, int]
    coords: List[Tuple[str, Tuple[Any, Dict[str, Any]]]] = field(default_factory=list)
    profile: str = "diagnostic"
    attrs: Dict[str, Any] = field(default_factory=dict)
    time_dtype: Any = np.float64

    def __post_init__(self) -> None:
        self.dims = [(str(k), int(v)) for k, v in self.dims]
        self.time_dim = str(self.time_dim)
        self.vars = [(str(k), v) for k, v in self.vars]
        self.chunk_shape = {str(k): int(v) for k, v in self.chunk_shape.items()}
        self.shard_shape = {str(k): int(v) for k, v in self.shard_shape.items()}
        self.coords = [
            (str(k), (np.asarray(vals), dict(a))) for k, (vals, a) in self.coords
        ]
        self.attrs = dict(self.attrs)
        self.time_dtype = np.dtype(self.time_dtype)

    @property
    def dim_lengths(self) -> Dict[str, int]:
        return {k: v for k, v in self.dims}


# --- the write handle -------------------------------------------------------- #


@dataclass
class ZarrWriteHandle:
    """Opaque handle threaded through :meth:`write_record` / :meth:`write_close`."""

    base: str
    schema: OutputSchema
    group: Any
    store: Any
    arrays: Dict[str, Any]
    codec: BloscProfile
    shard_time: int
    #: buffered (not-yet-flushed) time-coord values for the current shard
    time_buffer: List[float] = field(default_factory=list)
    #: var name -> list of spatial slabs buffered for the current shard
    buffers: Dict[str, List[np.ndarray]] = field(default_factory=dict)
    total_records: int = 0
    shard_time_index: int = 0
    time_shards: List[Dict[str, Any]] = field(default_factory=list)
    shard_t_start: Optional[float] = None
    last_t: Optional[float] = None


# --- dtype <-> Zarr v3 data_type (for the manifest fingerprint) -------------- #

_V3_DTYPE = {
    np.dtype("float64"): "float64",
    np.dtype("float32"): "float32",
    np.dtype("int32"): "int32",
    np.dtype("int64"): "int64",
    np.dtype("uint32"): "uint32",
    np.dtype("uint64"): "uint64",
}


def _v3_dtype(dt: np.dtype) -> str:
    dt = np.dtype(dt)
    if dt not in _V3_DTYPE:
        raise ValueError(f"unsupported output dtype {dt} for Zarr v3")
    return _V3_DTYPE[dt]


def _v3_fill(dt: np.dtype) -> Any:
    return 0.0 if np.issubdtype(np.dtype(dt), np.floating) else 0


# --- output base resolution -------------------------------------------------- #


def _open_output_store(base_url: str):
    """Resolve ``base_url`` to a writable zarr store.

    ``s3://…`` → ``FsspecStore`` (s3fs / the ``s3`` extra); ``file://…`` or a
    bare path → ``LocalStore``. Imported lazily so the base install needs neither
    ``zarr`` nor ``s3fs``.
    """
    from zarr.storage import LocalStore

    if base_url.startswith("s3://") or "://" in base_url and not base_url.startswith(
        "file://"
    ):
        # Object-store / any fsspec-addressable target (needs the ``s3`` extra).
        from zarr.storage import FsspecStore

        return FsspecStore.from_url(base_url, read_only=False), base_url
    if base_url.startswith("file://"):
        path = base_url[len("file://") :]
    else:
        path = base_url
    os.makedirs(path, exist_ok=True)
    return LocalStore(path), path


# --- the writer -------------------------------------------------------------- #


class ZarrWriter:
    """The ``zarr`` streaming writer (sharded Zarr v3). The write mirror of
    :class:`earthsciio.backends.zarr.ZarrReader`."""

    NAME = "zarr"

    def name(self) -> str:
        return self.NAME

    # -- write_open --------------------------------------------------------- #

    def write_open(self, base_url: str, schema: OutputSchema) -> ZarrWriteHandle:
        """Create the store: group node, static coords (written once), the
        growable time coordinate, and every streaming var at ``shape[time] = 0``.
        Returns the handle threaded through the rest of the lifecycle."""
        import zarr
        from zarr.codecs import BloscCodec

        codec = _profile(schema.profile)
        store, base = _open_output_store(base_url)

        # validate the chunk/shard grid
        for d, _ in schema.dims:
            if d not in schema.chunk_shape:
                raise ValueError(f"dim {d!r} missing from chunk_shape")
            if d not in schema.shard_shape:
                raise ValueError(f"dim {d!r} missing from shard_shape")
            if schema.shard_shape[d] % schema.chunk_shape[d] != 0:
                raise ValueError(
                    f"shard_shape[{d}]={schema.shard_shape[d]} is not a multiple of "
                    f"chunk_shape[{d}]={schema.chunk_shape[d]}"
                )
        shard_time = schema.shard_shape[schema.time_dim]

        group = zarr.open_group(store=store, mode="w", zarr_format=3)
        for k, v in schema.attrs.items():
            group.attrs[k] = v

        blosc = BloscCodec(
            cname=codec.cname,
            clevel=codec.clevel,
            shuffle="shuffle" if codec.shuffle else "noshuffle",
        )
        dimlen = schema.dim_lengths

        def _mk(name, dims, dtype, shape):
            chunks = tuple(schema.chunk_shape[d] for d in dims)
            shards = tuple(schema.shard_shape[d] for d in dims)
            return group.create_array(
                name=name,
                shape=tuple(shape),
                chunks=chunks,
                shards=shards,
                dtype=np.dtype(dtype),
                compressors=[blosc],
                dimension_names=list(dims),
                fill_value=_v3_fill(dtype),
                overwrite=True,
            )

        arrays: Dict[str, Any] = {}

        # static coords (values known now) — written once; time coord is separate
        time_attrs: Dict[str, Any] = {}
        for nm, (vals, attrs) in schema.coords:
            if nm == schema.time_dim:
                time_attrs = attrs
                continue
            varr = np.asarray(vals)
            a = _mk(nm, [nm], varr.dtype, varr.shape)
            a[...] = varr
            for k, v in attrs.items():
                a.attrs[k] = v
            arrays[nm] = a

        # growable time coordinate (1-D), starts empty
        tarr = _mk(schema.time_dim, [schema.time_dim], schema.time_dtype, (0,))
        for k, v in time_attrs.items():
            tarr.attrs[k] = v
        arrays[schema.time_dim] = tarr

        # streaming vars, shape[time] = 0
        for nm, ov in schema.vars:
            if schema.time_dim not in ov.dims:
                raise ValueError(
                    f"streaming var {nm!r} must include the time dim "
                    f"{schema.time_dim!r}"
                )
            shape = [0 if d == schema.time_dim else dimlen[d] for d in ov.dims]
            a = _mk(nm, ov.dims, ov.dtype, shape)
            for k, v in ov.attrs.items():
                a.attrs[k] = v
            arrays[nm] = a

        h = ZarrWriteHandle(
            base=base,
            schema=schema,
            group=group,
            store=store,
            arrays=arrays,
            codec=codec,
            shard_time=shard_time,
        )
        h.buffers = {nm: [] for nm, _ in schema.vars}
        return h

    # -- write_record ------------------------------------------------------- #

    def write_record(
        self, h: ZarrWriteHandle, t: float, arrays: Dict[str, Any], region=None
    ) -> None:
        """Buffer one time record. When the buffer fills a shard's worth of time
        steps (``shard_shape[time_dim]``) it is flushed durably."""
        if region is not None:
            raise ValueError("write_record `region` is reserved (None in Wave 4)")
        s = h.schema
        if not h.time_buffer:
            h.shard_t_start = float(t)
        h.time_buffer.append(float(t))
        for nm, ov in s.vars:
            if nm not in arrays:
                raise ValueError(f"write_record missing array for var {nm!r}")
            slab = np.asarray(arrays[nm], dtype=ov.dtype)
            expected = tuple(
                s.dim_lengths[d] for d in ov.dims if d != s.time_dim
            )
            if slab.shape != expected:
                raise ValueError(
                    f"var {nm!r}: slab shape {slab.shape} != expected {expected} "
                    f"(dims minus time)"
                )
            h.buffers[nm].append(slab)
        h.last_t = float(t)
        if len(h.time_buffer) >= h.shard_time:
            self._flush(h)

    # -- flush -------------------------------------------------------------- #

    def _flush(self, h: ZarrWriteHandle) -> None:
        n = len(h.time_buffer)
        if n == 0:
            return
        s = h.schema
        old = h.total_records
        new = old + n

        # time coordinate
        tarr = h.arrays[s.time_dim]
        tarr.resize((new,))
        tarr[old:new] = np.asarray(h.time_buffer, dtype=s.time_dtype)

        # streaming vars: resize the time axis and write the buffered slab block
        for nm, ov in s.vars:
            a = h.arrays[nm]
            ti = ov.dims.index(s.time_dim)
            newshape = list(a.shape)
            newshape[ti] = new
            a.resize(tuple(newshape))
            block = np.stack(h.buffers[nm], axis=ti)  # (…, n along time, …)
            sel = tuple(
                slice(old, new) if d == ti else slice(None)
                for d in range(len(ov.dims))
            )
            a[sel] = block

        h.time_shards.append(
            {
                "index": h.shard_time_index,
                "t_start": h.shard_t_start,
                "t_end": h.last_t,
                "n_records": n,
            }
        )
        h.total_records = new
        h.shard_time_index += 1
        h.time_buffer = []
        h.buffers = {nm: [] for nm, _ in s.vars}
        self._write_manifest(h)

    # -- write_flush (durable checkpoint barrier) --------------------------- #

    def write_flush(self, h: ZarrWriteHandle) -> None:
        """Force any buffered-but-unflushed records out as a durable (partial)
        shard and refresh the output manifest — the checkpoint durable barrier
        (RFC §10, §16.7). After this returns, a restart reading the manifest sees
        every record written so far as committed. No-op when the buffer is empty."""
        if h.time_buffer:
            self._flush(h)

    # -- write_close -------------------------------------------------------- #

    def write_close(self, h: ZarrWriteHandle) -> Dict[str, Any]:
        """Flush any trailing partial shard, finalize, and write the end-of-run
        output manifest (the close record). Returns the manifest dict."""
        if h.time_buffer:
            self._flush(h)
        return self._write_manifest(h)

    # -- output manifest ---------------------------------------------------- #

    def _manifest_dict(self, h: ZarrWriteHandle) -> Dict[str, Any]:
        from datetime import datetime, timezone

        s = h.schema
        codec = {
            "id": "blosc",
            "cname": h.codec.cname,
            "clevel": h.codec.clevel,
            "shuffle": "shuffle" if h.codec.shuffle else "noshuffle",
        }
        vars_meta = [
            {"name": nm, "dims": list(ov.dims), "dtype": _v3_dtype(ov.dtype)}
            for nm, ov in s.vars
        ]
        return {
            "schema": OUTPUT_MANIFEST_SCHEMA,
            "base_url": h.base,
            "format": "zarr",
            "zarr_format": 3,
            "profile": s.profile,
            "codec": codec,
            "time_dim": s.time_dim,
            "dims": [[k, v] for k, v in s.dims],
            "vars": vars_meta,
            "chunk_shape": dict(s.chunk_shape),
            "shard_shape": dict(s.shard_shape),
            "time_shards": list(h.time_shards),
            "last_t": h.last_t,
            "total_records": h.total_records,
            "written_at": datetime.now(timezone.utc)
            .replace(microsecond=0)
            .strftime("%Y-%m-%dT%H:%M:%SZ"),
        }

    def _write_manifest(self, h: ZarrWriteHandle) -> Dict[str, Any]:
        """Write ``output_manifest.json`` into the store as a sibling object,
        through the same store the arrays use (so it lands on local FS or S3)."""
        m = self._manifest_dict(h)
        text = json.dumps(m, indent=2, sort_keys=True) + "\n"
        self._put_manifest_bytes(h, text.encode("utf-8"))
        return m

    @staticmethod
    def _put_manifest_bytes(h: ZarrWriteHandle, data: bytes) -> None:
        from zarr.core.buffer import default_buffer_prototype
        from zarr.core.sync import sync

        buf = default_buffer_prototype().buffer.from_bytes(data)
        sync(h.store.set("output_manifest.json", buf))

File: rust/src/format/zarr_write.rs
//! The `zarr` **writer** — a sharded Zarr **v3** output backend built on the
//! mainstream [`zarrs`] crate (the write mirror of [`ZarrReader`](super::ZarrReader)).
//!
//! It emits the same store shape the Julia reference writer does
//! (`julia/src/zarr_write.jl`, streaming-output-sinks RFC §16), so a store written
//! by any track is cross-readable by the others:
//!
//!   * `zarr_format: 3`, one `zarr.json` per group + per array.
//!   * The **`sharding_indexed`** codec: the array's `chunk_grid.chunk_shape` is the
//!     SHARD (outer, write) shape; the sharding codec's `chunk_shape` is the inner
//!     (read) chunk. Inner pipeline `[bytes(little), blosc(zstd+shuffle)]`; index
//!     pipeline `[bytes(little), crc32c]`, `index_location: "end"`.
//!   * `dimension_names` + a mirrored CF `_ARRAY_DIMENSIONS` attribute; per-array
//!     `attributes` carry the caller's CF variable attributes; static coordinate
//!     arrays are written with values + attrs.
//!   * A per-store JSON output manifest (`output_manifest.json`, schema
//!     `earthsciio/output-manifest/v1`) mirroring `EarthSciIO`'s Julia writer.
//!
//! Unlike the Julia writer's incremental `write_record!` streaming loop, this Rust
//! writer commits a whole dataset in one [`write_zarr_v3`] call (the streaming
//! record/flush handle is future work). `zarrs` owns the codec pipeline (blosc
//! encode, crc32c, shard assembly), so the crate's own code stays
//! `#![forbid(unsafe_code)]`.
//!
//! **Tolerance policy.** Cross-language output is NOT byte-identical (Blosc
//! container bytes are host/version dependent). Conformance is on **decoded
//! arrays** within tolerance plus metadata equality (RFC §16).
//!
//! Local output uses `zarrs`' pure filesystem store; an S3 (`s3://`) target is
//! handled by the `object_store`-backed opener (feature `object-store`).

use std::collections::BTreeMap;
use std::path::Path;
use std::sync::Arc;

use serde_json::{json, Map, Value};
use zarrs::array::{Array, ArrayMetadata};
use zarrs::filesystem::FilesystemStore;

use crate::error::{Error, Result};

/// A pinned Blosc codec profile: `cname` / `clevel` / byte-`shuffle`
/// (RFC §16). Kept as constants so the Julia/Python/Rust ports agree on params.
#[derive(Debug, Clone, Copy)]
pub struct BloscProfile {
    /// Blosc compressor name (e.g. `"zstd"`).
    pub cname: &'static str,
    /// Blosc compression level.
    pub clevel: u8,
    /// Byte-shuffle on/off.
    pub shuffle: bool,
}

/// Diagnostic profile — Blosc **zstd** + byte-shuffle, moderate level (5).
pub const BLOSC_DIAGNOSTIC: BloscProfile = BloscProfile {
    cname: "zstd",
    clevel: 5,
    shuffle: true,
};
/// Checkpoint profile — **lossless** Blosc zstd (level 7) + byte-shuffle.
pub const BLOSC_CHECKPOINT: BloscProfile = BloscProfile {
    cname: "zstd",
    clevel: 7,
    shuffle: true,
};

fn profile(name: &str) -> Result<BloscProfile> {
    match name {
        "diagnostic" => Ok(BLOSC_DIAGNOSTIC),
        "checkpoint" => Ok(BLOSC_CHECKPOINT),
        other => Err(err(format!(
            "unknown codec profile '{other}' (expected 'diagnostic' or 'checkpoint')"
        ))),
    }
}

/// One static (time-independent) coordinate array: 1-D over its own dim `name`,
/// with `values` and CF `attrs` (e.g. `units`, `standard_name`). `float64`.
#[derive(Debug, Clone)]
pub struct WriteCoord {
    /// Coordinate (and dimension) name.
    pub name: String,
    /// The coordinate values (1-D).
    pub values: Vec<f64>,
    /// CF attributes written into the array node's `attributes`.
    pub attrs: Map<String, Value>,
}

/// One streaming output variable: its on-disk `dims` (file order), CF `attrs`,
/// and its `data` flattened **row-major (C order)** over the shape implied by
/// `dims`. `float64`.
#[derive(Debug, Clone)]
pub struct WriteVar {
    /// Variable name.
    pub name: String,
    /// Ordered dimension names (must all be schema dims; includes `time_dim`).
    pub dims: Vec<String>,
    /// CF attributes (e.g. `units`, `coordinates`).
    pub attrs: Map<String, Value>,
    /// Row-major (C-order) values over `dims`.
    pub data: Vec<f64>,
}

/// The output schema (a whole-dataset write). Mirrors the Julia `OutputSchema`.
#[derive(Debug, Clone)]
pub struct OutputSchema {
    /// ORDERED dim name => length (the `time_dim` entry's length is the number of
    /// records written).
    pub dims: Vec<(String, usize)>,
    /// The growable/record axis name.
    pub time_dim: String,
    /// dim name => INNER chunk length (every dim + coord name needs an entry).
    pub chunk_shape: BTreeMap<String, usize>,
    /// dim name => SHARD length (a multiple of the inner chunk length per dim).
    pub shard_shape: BTreeMap<String, usize>,
    /// ORDERED static coordinate arrays (written once, with values).
    pub coords: Vec<WriteCoord>,
    /// ORDERED streaming variables.
    pub vars: Vec<WriteVar>,
    /// Group-level attributes.
    pub group_attrs: Map<String, Value>,
    /// Codec profile name: `"diagnostic"` or `"checkpoint"`.
    pub profile: String,
}

fn err(detail: impl Into<String>) -> Error {
    Error::Format {
        format: "zarr".to_string(),
        detail: detail.into(),
    }
}

/// Write a complete sharded Zarr **v3** store at `base` (a local directory)
/// following [`OutputSchema`], plus the JSON output manifest. `zarrs` performs
/// all codec work (blosc, crc32c, shard packing).
///
/// # Errors
/// Returns [`Error::Format`] on schema inconsistency (e.g. a shard shape not a
/// multiple of the inner chunk), or [`Error::Io`] / a `zarrs` error on write.
pub fn write_zarr_v3(base: &Path, schema: &OutputSchema) -> Result<()> {
    let codec = profile(&schema.profile)?;
    let dim_len: BTreeMap<&str, usize> =
        schema.dims.iter().map(|(k, v)| (k.as_str(), *v)).collect();

    // Validate the chunk/shard grid.
    for (d, _) in &schema.dims {
        let c = *schema
            .chunk_shape
            .get(d)
            .ok_or_else(|| err(format!("dim '{d}' missing from chunk_shape")))?;
        let s = *schema
            .shard_shape
            .get(d)
            .ok_or_else(|| err(format!("dim '{d}' missing from shard_shape")))?;
        if c == 0 || s == 0 || s % c != 0 {
            return Err(err(format!(
                "shard_shape[{d}]={s} must be a nonzero multiple of chunk_shape[{d}]={c}"
            )));
        }
    }

    std::fs::create_dir_all(base).map_err(|e| Error::io(Some(base.to_path_buf()), e))?;
    let store = Arc::new(
        FilesystemStore::new(base)
            .map_err(|e| err(format!("open filesystem store at {}: {e}", base.display())))?,
    );

    // Group metadata (zarr.json at the root).
    let group_meta = json!({
        "zarr_format": 3,
        "node_type": "group",
        "attributes": Value::Object(schema.group_attrs.clone()),
    });
    write_group_json(base, &group_meta)?;

    // Static coordinate arrays (values known now).
    for co in &schema.coords {
        let shape = vec![co.values.len()];
        write_array(
            store.clone(),
            &co.name,
            std::slice::from_ref(&co.name),
            &shape,
            schema,
            &codec,
            &co.attrs,
            &co.values,
        )?;
    }

    // Streaming variables.
    for v in &schema.vars {
        if !v.dims.iter().any(|d| d == &schema.time_dim) {
            return Err(err(format!(
                "streaming var '{}' must include the time dim '{}'",
                v.name, schema.time_dim
            )));
        }
        let shape: Vec<usize> = v
            .dims
            .iter()
            .map(|d| {
                dim_len
                    .get(d.as_str())
                    .copied()
                    .ok_or_else(|| err(format!("var '{}' dim '{d}' not in schema dims", v.name)))
            })
            .collect::<Result<_>>()?;
        let expect: usize = shape.iter().product();
        if v.data.len() != expect {
            return Err(err(format!(
                "var '{}' data length {} != product(shape {:?}) = {expect}",
                v.name,
                v.data.len(),
                shape
            )));
        }
        write_array(
            store.clone(),
            &v.name,
            &v.dims,
            &shape,
            schema,
            &codec,
            &v.attrs,
            &v.data,
        )?;
    }

    write_output_manifest(base, schema, &codec)?;
    Ok(())
}

/// The `sharding_indexed` codec dict for inner chunk shape `inner` (RFC §16).
fn sharding_codec(inner: &[usize], codec: &BloscProfile, typesize: usize) -> Value {
    json!({
        "name": "sharding_indexed",
        "configuration": {
            "chunk_shape": inner,
            "codecs": [
                {"name": "bytes", "configuration": {"endian": "little"}},
                {"name": "blosc", "configuration": {
                    "cname": codec.cname,
                    "clevel": codec.clevel,
                    "shuffle": if codec.shuffle { "shuffle" } else { "noshuffle" },
                    "typesize": typesize,
                    "blocksize": 0,
                }},
            ],
            "index_codecs": [
                {"name": "bytes", "configuration": {"endian": "little"}},
                {"name": "crc32c"},
            ],
            "index_location": "end",
        },
    })
}

/// Build the v3 array metadata dict (mirrors Julia `_array_meta_dict`).
fn array_meta(
    dims: &[String],
    shape: &[usize],
    schema: &OutputSchema,
    codec: &BloscProfile,
    attrs: &Map<String, Value>,
) -> Result<Value> {
    let inner: Vec<usize> = dims
        .iter()
        .map(|d| {
            schema
                .chunk_shape
                .get(d)
                .copied()
                .ok_or_else(|| err(format!("dim '{d}' missing from chunk_shape")))
        })
        .collect::<Result<_>>()?;
    let shard: Vec<usize> = dims
        .iter()
        .map(|d| {
            schema
                .shard_shape
                .get(d)
                .copied()
                .ok_or_else(|| err(format!("dim '{d}' missing from shard_shape")))
        })
        .collect::<Result<_>>()?;
    let mut a = attrs.clone();
    a.insert("_ARRAY_DIMENSIONS".to_string(), json!(dims));
    Ok(json!({
        "zarr_format": 3,
        "node_type": "array",
        "shape": shape,
        "data_type": "float64",
        "chunk_grid": {"name": "regular", "configuration": {"chunk_shape": shard}},
        "chunk_key_encoding": {"name": "default", "configuration": {"separator": "/"}},
        "fill_value": 0.0,
        "codecs": [sharding_codec(&inner, codec, std::mem::size_of::<f64>())],
        "attributes": Value::Object(a),
        "dimension_names": dims,
    }))
}

/// Create the array node (writing its `zarr.json`) and store its `data`.
#[allow(clippy::too_many_arguments)]
fn write_array(
    store: Arc<FilesystemStore>,
    name: &str,
    dims: &[String],
    shape: &[usize],
    schema: &OutputSchema,
    codec: &BloscProfile,
    attrs: &Map<String, Value>,
    data: &[f64],
) -> Result<()> {
    let meta_json = array_meta(dims, shape, schema, codec, attrs)?;
    let metadata: ArrayMetadata = serde_json::from_value(meta_json)
        .map_err(|e| err(format!("array '{name}' metadata is not valid v3: {e}")))?;
    let array = Array::new_with_metadata(store, &format!("/{name}"), metadata)
        .map_err(|e| err(format!("build zarr array '{name}': {e}")))?;
    array
        .store_metadata()
        .map_err(|e| err(format!("write metadata for '{name}': {e}")))?;
    // subset_all is the whole array; where the shard grid divides the shape evenly
    // this writes each shard exactly once (no read-modify-write).
    let subset = array.subset_all();
    array
        .store_array_subset(&subset, data)
        .map_err(|e| err(format!("write data for '{name}': {e}")))?;
    Ok(())
}

/// Write the group root `zarr.json` (pretty JSON, atomic via a sibling temp).
fn write_group_json(base: &Path, meta: &Value) -> Result<()> {
    let bytes = serde_json::to_vec_pretty(meta)
        .map_err(|e| err(format!("serialize group metadata: {e}")))?;
    let path = base.join("zarr.json");
    atomic_write(&path, &bytes)
}

/// Write the output manifest (`earthsciio/output-manifest/v1`) mirroring the
/// Julia `OutputManifest` fields.
fn write_output_manifest(base: &Path, schema: &OutputSchema, codec: &BloscProfile) -> Result<()> {
    let n_records = schema
        .dims
        .iter()
        .find(|(d, _)| d == &schema.time_dim)
        .map(|(_, l)| *l)
        .unwrap_or(0);
    let shard_time = schema
        .shard_shape
        .get(&schema.time_dim)
        .copied()
        .unwrap_or(n_records.max(1));

    // The time coordinate values (if a coord for the time dim was supplied) give
    // the per-shard t_start/t_end; otherwise fall back to record indices.
    let time_vals: Option<&Vec<f64>> = schema
        .coords
        .iter()
        .find(|c| c.name == schema.time_dim)
        .map(|c| &c.values);

    let mut time_shards = Vec::new();
    if shard_time > 0 && n_records > 0 {
        let mut index = 0usize;
        let mut start = 0usize;
        while start < n_records {
            let end = (start + shard_time).min(n_records);
            let n = end - start;
            let t_start = time_vals.and_then(|v| v.get(start).copied()).unwrap_or(start as f64);
            let t_end = time_vals
                .and_then(|v| v.get(end - 1).copied())
                .unwrap_or((end - 1) as f64);
            time_shards.push(json!({
                "index": index, "t_start": t_start, "t_end": t_end, "n": n,
            }));
            index += 1;
            start = end;
        }
    }
    let last_t: Value = if n_records == 0 {
        Value::Null
    } else {
        time_vals
            .and_then(|v| v.last().copied())
            .map(|t| json!(t))
            .unwrap_or_else(|| json!((n_records - 1) as f64))
    };

    let vars: Vec<Value> = schema
        .vars
        .iter()
        .map(|v| json!({"name": v.name, "dims": v.dims, "dtype": "float64"}))
        .collect();
    let dims: Vec<Value> = schema
        .dims
        .iter()
        .map(|(k, v)| json!({"name": k, "length": v}))
        .collect();

    let manifest = json!({
        "schema": "earthsciio/output-manifest/v1",
        "base_url": base.to_string_lossy(),
        "format": "zarr",
        "zarr_format": 3,
        "profile": schema.profile,
        "codec": {
            "id": "blosc",
            "cname": codec.cname,
            "clevel": codec.clevel,
            "shuffle": if codec.shuffle { "shuffle" } else { "noshuffle" },
        },
        "time_dim": schema.time_dim,
        "dims": dims,
        "vars": vars,
        "chunk_shape": schema.chunk_shape,
        "shard_shape": schema.shard_shape,
        "time_shards": time_shards,
        "last_t": last_t,
        "n_records": n_records,
        "created_at": crate::clock::now_rfc3339(),
    });
    let bytes = serde_json::to_vec_pretty(&manifest)
        .map_err(|e| err(format!("serialize output manifest: {e}")))?;
    atomic_write(&base.join("output_manifest.json"), &bytes)
}

/// Write `bytes` to `path` atomically via a sibling temp + rename.
fn atomic_write(path: &Path, bytes: &[u8]) -> Result<()> {
    let dir = path.parent().unwrap_or_else(|| Path::new("."));
    let mut tmp = tempfile::Builder::new()
        .prefix(".zarr-")
        .suffix(".part")
        .tempfile_in(dir)
        .map_err(|e| Error::io(Some(dir.to_path_buf()), e))?;
    use std::io::Write as _;
    tmp.write_all(bytes)
        .map_err(|e| Error::io(Some(tmp.path().to_path_buf()), e))?;
    tmp.flush()
        .map_err(|e| Error::io(Some(tmp.path().to_path_buf()), e))?;
    tmp.persist(path)
        .map_err(|e| Error::io(Some(path.to_path_buf()), e.error))?;
    Ok(())
}
